shared-module/audiofilters/Filter.c

```diff
@@ -0,0 +1,346 @@
+            if (!self->more_data) { // The sample has indicated it has no more data to play
+                if (self->loop && self->sample) { // If we are supposed to loop reset the sample to the start
+                    audiosample_reset_buffer(self->sample, false, 0);
+                } else { // If we were not supposed to loop the sample, stop playing it
+                    self->sample = NULL;
+                }
+            }
+            if (self->sample) {
+                // Load another sample buffer to play
+                audioio_get_buffer_result_t result = audiosample_get_buffer(self->sample, false, 0, (uint8_t **)&self->sample_remaining_buffer, &self->sample_buffer_length);
+                // Track length in terms of words.
+                self->sample_buffer_length /= (self->bits_per_sample / 8);
+                self->more_data = result == GET_BUFFER_MORE_DATA;
+            }
+        }
+
+        // If we have a sample, filter it
+        if (self->sample != NULL) {
+            // Determine how many bytes we can process to our buffer, the less of the sample we have left and our buffer remaining
+            uint32_t n = MIN(self->sample_buffer_length, length);
+
+            int16_t *sample_src = (int16_t *)self->sample_remaining_buffer; // for 16-bit samples
+            int8_t *sample_hsrc = (int8_t *)self->sample_remaining_buffer; // for 8-bit samples
+
+            if (mix <= 0.01 || self->filter_obj == mp_const_none) { // if mix is zero pure sample only or no biquad filter object is provided
+                for (uint32_t i = 0; i < n; i++) {
+                    if (MP_LIKELY(self->bits_per_sample == 16)) {
+                        word_buffer[i] = sample_src[i];
+                    } else {
+                        hword_buffer[i] = sample_hsrc[i];
+                    }
+                }
+            } else {
+                uint32_t i = 0;
+                while (i < n) {
+                    uint32_t n_samples = MIN(SYNTHIO_MAX_DUR, n - i);
+
+                    // Fill filter buffer with samples
+                    for (uint32_t j = 0; j < n_samples; j++) {
+                        if (MP_LIKELY(self->bits_per_sample == 16)) {
+                            self->filter_buffer[0][j] = sample_src[i + j];
+                        } else {
+                            if (self->samples_signed) {
+                                self->filter_buffer[0][j] = sample_hsrc[i + j];
+                            } else {
+                                // Be careful here changing from an 8 bit unsigned to signed into a 32-bit signed
+                                self->filter_buffer[0][j] = (int8_t)(((uint8_t)sample_hsrc[i + j]) ^ 0x80);
+                            }
+                        }
+                    }
+
+                    // Copy original signal for mixing back in later
+                    memcpy(self->filter_buffer[1], self->filter_buffer[0], n_samples);
```

Comment:
Is this just to make the mixing easier, or is there a reason to not just use the original samples on line 296 to do the mix?

Reply:
I think you have a point here. The second filter buffer (which I was just using as a copy for mixing later) isn't exactly necessary if we can pull from the original sample again in the mix loop.

Reply:
I've updated the module with this change and it performs as expected.